### ISSUE TITLE
Add two-step merge testing for Android

### ISSUE THREAD

**Issue by inferno-chromium** [google/clusterfuzz#1732]
- Need to rebuild android test binaries with revision > https://github.com/llvm/llvm-project/commit/f054067
- Testing inside fuzz that two step merge failed.

See https://github.com/google/clusterfuzz/pull/1731

**Comment by Dor1s** [comment]
Who can re-build those binaries? Should we ask Mitch or Hamzeh? I'm pretty sure that the actual Android builds are newer than that, it's just our test binaries in `src/python/tests/core/bot/fuzzers/libFuzzer/data`

**Comment by inferno-chromium** [comment]
> Who can re-build those binaries? Should we ask Mitch or Hamzeh? I'm pretty sure that the actual Android builds are newer than that, it's just our test binaries in `src/python/tests/core/bot/fuzzers/libFuzzer/data`

Yes either of them should be fine (i did it before, but dont have cycles right now). And yes it just uses the same fuzzers from data dir.

**Comment by Dor1s** [comment]
@hctim could you please help us to update test binaries for Android?

We need four test fuzzers:
1) https://github.com/google/clusterfuzz/blob/master/src/python/tests/core/bot/fuzzers/libFuzzer/data/test_fuzzer.cc
2) https://github.com/google/clusterfuzz/blob/master/src/python/tests/core/bot/fuzzers/libFuzzer/data/crash_with_A_fuzzer.cc
3) https://github.com/google/clusterfuzz/blob/master/src/python/tests/core/bot/fuzzers/libFuzzer/data/analyze_dict_fuzzer.cc
4) https://github.com/google/clusterfuzz/blob/master/src/python/tests/core/bot/fuzzers/libFuzzer/data/always_crash_fuzzer.cc

The third one has the flags we used sometime in the past:

```
// clang++ -g -fsanitize=address -fsanitize-coverage=trace-pc-guard \
// -O0 -std=c++11 analyze_dict_fuzzer.cc libFuzzer.a -o analyze_dict_fuzzer
```

but I'm pretty sure that `-fsanitize=address,fuzzer` would also work, we just need it to be for Android.

Hopefully that wouldn't take more than a couple minutes for you!

**Comment by hctim** [comment]
From offline conversation, SDK is at:

```
commit b777009ceddf856ff9614701a706da97785a6162
Author: Craig Topper <craig.topper@intel.com>
Date:   Thu Jul 11 01:01:39 2019 +0000```
```

Toolchain is at:

```
commit dfd6374c784fe9e6d6863c659f8734b91d6fe52a
Author: Kadir Cetinkaya <kadircet@google.com>
Date:   Fri Sep 6 09:49:40 2019 +0200
```

(both prior to the patches). mmoroz@ says no rush on getting new binaries, so we can just wait until the next toolchain roll happens. Can you ping me in a month or so? There should be a new toolchain by then.

**Comment by Dor1s** [comment]
Yes, will do. Thank you, Mitch!

**Comment by hctim** [comment]
Got you some binaries at:
```
commit e09107ab80dced55414fa458cf78e6cdfe90da6e (HEAD -> master, origin/master, origin/HEAD)
Author: Raul Tambre <raul@tambre.ee>
Date:   Fri Sep 4 19:10:09 2020 +0300

    [Sema] Introduce BuiltinAttr, per-declaration builtin-ness
```

(hope this isn't *too* new)

Quick clarification would be great:
 - Is HWASan okay (i.e. `-fsanitize=hwaddress,fuzzer`)?
 - Any other build flags you need for these binaries?

I've pushed https://github.com/hctim/clusterfuzz/blob/android_bins/oss-fuzz.tar to my own branch for you to take a look at. They were built with `clang++ -fPIC -fsanitize=fuzzer,<variant> %s -o %s.o` and I've also included libcxx and the sanitizer runtimes.

Bear in mind that Android installs its own signal handlers, so:
 
 - ASan binaries dump stacktraces/testcases out of the box (but we don't want to support ASan any more, so HWASan would be preferred).
 - HWASan binaries don't install signal handlers unless you ask it to: `HWASAN_OPTIONS=handle_segv=1`, etc.
 - no sanitizer binaries install their signal handlers early, and then the platform clobbers them. libFuzzer's handlers don't have the logic that the sanitizers do, so you can't get libFuzzer's handlers to dump the testcase :(.

Also bear in mind that I haven't given you a HWASan-ified libc, so only basic tests will work. Any heap-related stuff won't work as HWASan doesn't do interposition.

LMK if you need any changes. Works on my device, but worth giving it a crack on yours.

**Comment by Dor1s** [comment]
Thanks so much, @hctim!
